// @shopify/polaris/dist/esm/components/OptionList/components/Checkbox/Checkbox.js
import React$1, { useState } from 'react';
import { useUniqueId } from '../../../../utilities/unique-id/hooks.js';
import { classNames } from '../../../../utilities/css.js';
import { TickSmallMinor } from '@shopify/polaris-icons';
import { Icon as Icon$1 } from '../../../Icon/Icon.js';
import styles from './Checkbox.scss.js';

var _ref = /*#__PURE__*/React$1.createElement(Icon$1, {
  source: TickSmallMinor
});

function Checkbox({
  id: idProp,
  checked = false,
  disabled,
  active,
  onChange,
  name,
  value,
  role
}) {
  const id = useUniqueId('Checkbox', idProp);
  const [keyFocused, setKeyFocused] = useState(false);
  const className = classNames(styles.Checkbox, active && styles.active);

  const handleBlur = () => {
    setKeyFocused(false);
  };

  const handleKeyUp = () => {
    !keyFocused && setKeyFocused(true);
  };

  const inputClassName = classNames(styles.Input, keyFocused && styles.keyFocused);
  return /*#__PURE__*/React$1.createElement("div", {
    className: className
  }, /*#__PURE__*/React$1.createElement("input", {
    id: id,
    name: name,
    value: value,
    type: "checkbox",
    checked: checked,
    disabled: disabled,
    className: inputClassName,
    "aria-checked": checked,
    onChange: onChange,
    onBlur: handleBlur,
    onKeyUp: handleKeyUp,
    role: role
  }), /*#__PURE__*/React$1.createElement("div", {
    className: styles.Backdrop
  }), /*#__PURE__*/React$1.createElement("div", {
    className: styles.Icon
  }, _ref));
}

export { Checkbox };
