// @shopify/polaris/dist/esm/components/ActionMenu/components/MenuGroup/MenuGroup.js
import React$1, { useCallback } from 'react';
import { Popover as Popover$1 } from '../../../Popover/Popover.js';
import { ActionList as ActionList$1 } from '../../../ActionList/ActionList.js';
import { SecondaryAction as SecondaryAction$1 } from '../SecondaryAction/SecondaryAction.js';
import styles from './MenuGroup.scss.js';

function MenuGroup({
  accessibilityLabel,
  active,
  actions,
  details,
  title,
  icon,
  onClose,
  onOpen,
  getOffsetWidth
}) {
  const handleClose = useCallback(() => {
    onClose(title);
  }, [onClose, title]);
  const handleOpen = useCallback(() => {
    onOpen(title);
  }, [onOpen, title]);
  const handleOffsetWidth = useCallback(width => {
    if (!getOffsetWidth) return;
    getOffsetWidth(width);
  }, [getOffsetWidth]);
  const popoverActivator = /*#__PURE__*/React$1.createElement(SecondaryAction$1, {
    disclosure: true,
    icon: icon,
    accessibilityLabel: accessibilityLabel,
    onClick: handleOpen,
    getOffsetWidth: handleOffsetWidth
  }, title);
  return /*#__PURE__*/React$1.createElement(Popover$1, {
    active: Boolean(active),
    activator: popoverActivator,
    preferredAlignment: "left",
    onClose: handleClose,
    hideOnPrint: true
  }, /*#__PURE__*/React$1.createElement(ActionList$1, {
    items: actions,
    onActionAnyItem: handleClose
  }), details && /*#__PURE__*/React$1.createElement("div", {
    className: styles.Details
  }, details));
}

export { MenuGroup };
