// @shopify/polaris/dist/esm/components/Tag/Tag.js
import React$1 from 'react';
import { useI18n } from '../../utilities/i18n/hooks.js';
import { classNames } from '../../utilities/css.js';
import { CancelSmallMinor } from '@shopify/polaris-icons';
import { handleMouseUpByBlurring } from '../../utilities/focus.js';
import { Icon as Icon$1 } from '../Icon/Icon.js';
import styles from './Tag.scss.js';

var _ref = /*#__PURE__*/React$1.createElement(Icon$1, {
  source: CancelSmallMinor
});

function Tag({
  children,
  disabled = false,
  onClick,
  onRemove
}) {
  const i18n = useI18n();
  const className = classNames(styles.Tag, disabled && styles.disabled, onClick && styles.clickable, onRemove && styles.removable);
  const ariaLabel = i18n.translate('Polaris.Tag.ariaLabel', {
    children: children || ''
  });
  const removeButton = onRemove ? /*#__PURE__*/React$1.createElement("button", {
    type: "button",
    "aria-label": ariaLabel,
    className: styles.Button,
    onClick: onRemove,
    onMouseUp: handleMouseUpByBlurring,
    disabled: disabled
  }, _ref) : null;
  const tagMarkup = onClick ? /*#__PURE__*/React$1.createElement("button", {
    type: "button",
    disabled: disabled,
    className: className,
    onClick: onClick
  }, children) : /*#__PURE__*/React$1.createElement("span", {
    className: className
  }, /*#__PURE__*/React$1.createElement("span", {
    title: children,
    className: styles.TagText
  }, children), removeButton);
  return tagMarkup;
}

export { Tag };
