// @shopify/polaris/dist/esm/components/TopBar/components/Menu/Menu.js
import React$1 from 'react';
import { Popover as Popover$1 } from '../../../Popover/Popover.js';
import { ActionList as ActionList$1 } from '../../../ActionList/ActionList.js';
import { Message as Message$1 } from './components/Message/Message.js';
import styles from './Menu.scss.js';

function Menu(props) {
  const {
    actions,
    onOpen,
    onClose,
    open,
    activatorContent,
    message,
    colorScheme,
    accessibilityLabel
  } = props;
  const badgeProps = message && message.badge && {
    content: message.badge.content,
    status: message.badge.status
  };
  const messageMarkup = message && /*#__PURE__*/React$1.createElement(Message$1, {
    title: message.title,
    description: message.description,
    action: {
      onClick: message.action.onClick,
      content: message.action.content
    },
    link: {
      to: message.link.to,
      content: message.link.content
    },
    badge: badgeProps
  });
  const isFullHeight = Boolean(message);
  return /*#__PURE__*/React$1.createElement(Popover$1, {
    activator: /*#__PURE__*/React$1.createElement("div", {
      className: styles.ActivatorWrapper
    }, /*#__PURE__*/React$1.createElement("button", {
      type: "button",
      className: styles.Activator,
      onClick: onOpen,
      "aria-label": accessibilityLabel
    }, activatorContent)),
    active: open,
    onClose: onClose,
    fixed: true,
    fullHeight: isFullHeight,
    preferredAlignment: "right",
    colorScheme: colorScheme
  }, /*#__PURE__*/React$1.createElement(ActionList$1, {
    onActionAnyItem: onClose,
    sections: actions
  }), messageMarkup);
}

export { Menu };
