// @shopify/polaris/dist/esm/components/Frame/components/ContextualSaveBar/ContextualSaveBar.js
import React$1, { useCallback } from 'react';
import { useTheme } from '../../../../utilities/theme/hooks.js';
import { ThemeProvider as ThemeProvider$1 } from '../../../ThemeProvider/ThemeProvider.js';
import { useI18n } from '../../../../utilities/i18n/hooks.js';
import { classNames } from '../../../../utilities/css.js';
import { Image as Image$1 } from '../../../Image/Image.js';
import { Button as Button$1 } from '../../../Button/Button.js';
import { useToggle as useToggle$1 } from '../../../../utilities/use-toggle.js';
import { Stack as Stack$1 } from '../../../Stack/Stack.js';
import { getWidth as getWidth$1 } from '../../../../utilities/get-width.js';
import { DiscardConfirmationModal as DiscardConfirmationModal$1 } from './components/DiscardConfirmationModal/DiscardConfirmationModal.js';
import styles from './ContextualSaveBar.scss.js';

function ContextualSaveBar({
  alignContentFlush,
  message,
  saveAction,
  discardAction,
  fullWidth,
  contextControl
}) {
  const i18n = useI18n();
  const {
    logo
  } = useTheme();
  const {
    value: discardConfirmationModalVisible,
    toggle: toggleDiscardConfirmationModal,
    setFalse: closeDiscardConfirmationModal
  } = useToggle$1(false);
  const handleDiscardAction = useCallback(() => {
    if (discardAction && discardAction.onAction) {
      discardAction.onAction();
    }

    closeDiscardConfirmationModal();
  }, [closeDiscardConfirmationModal, discardAction]);
  const discardActionContent = discardAction && discardAction.content ? discardAction.content : i18n.translate('Polaris.ContextualSaveBar.discard');
  let discardActionHandler;

  if (discardAction && discardAction.discardConfirmationModal) {
    discardActionHandler = toggleDiscardConfirmationModal;
  } else if (discardAction) {
    discardActionHandler = discardAction.onAction;
  }

  const discardConfirmationModalMarkup = discardAction && discardAction.onAction && discardAction.discardConfirmationModal && /*#__PURE__*/React$1.createElement(DiscardConfirmationModal$1, {
    open: discardConfirmationModalVisible,
    onCancel: toggleDiscardConfirmationModal,
    onDiscard: handleDiscardAction
  });
  const discardActionMarkup = discardAction && /*#__PURE__*/React$1.createElement(Button$1, {
    url: discardAction.url,
    onClick: discardActionHandler,
    loading: discardAction.loading,
    disabled: discardAction.disabled,
    accessibilityLabel: discardAction.content
  }, discardActionContent);
  const saveActionContent = saveAction && saveAction.content ? saveAction.content : i18n.translate('Polaris.ContextualSaveBar.save');
  const saveActionMarkup = saveAction && /*#__PURE__*/React$1.createElement(Button$1, {
    primary: true,
    url: saveAction.url,
    onClick: saveAction.onAction,
    loading: saveAction.loading,
    disabled: saveAction.disabled,
    accessibilityLabel: saveAction.content
  }, saveActionContent);
  const width = getWidth$1(logo, 104);
  const imageMarkup = logo && /*#__PURE__*/React$1.createElement(Image$1, {
    style: {
      width
    },
    source: logo.contextualSaveBarSource || '',
    alt: ""
  });
  const logoMarkup = alignContentFlush || contextControl ? null : /*#__PURE__*/React$1.createElement("div", {
    className: styles.LogoContainer,
    style: {
      width
    }
  }, imageMarkup);
  const contextControlMarkup = contextControl ? /*#__PURE__*/React$1.createElement("div", {
    className: styles.ContextControl
  }, contextControl) : null;
  const contentsClassName = classNames(styles.Contents, fullWidth && styles.fullWidth);
  return /*#__PURE__*/React$1.createElement(React$1.Fragment, null, /*#__PURE__*/React$1.createElement(ThemeProvider$1, {
    theme: {
      colorScheme: 'inverse'
    }
  }, /*#__PURE__*/React$1.createElement("div", {
    className: styles.ContextualSaveBar
  }, contextControlMarkup, logoMarkup, /*#__PURE__*/React$1.createElement("div", {
    className: contentsClassName
  }, /*#__PURE__*/React$1.createElement("h2", {
    className: styles.Message
  }, message), /*#__PURE__*/React$1.createElement("div", {
    className: styles.ActionContainer
  }, /*#__PURE__*/React$1.createElement(Stack$1, {
    spacing: "tight",
    wrap: false
  }, discardActionMarkup, saveActionMarkup))))), discardConfirmationModalMarkup);
}

export { ContextualSaveBar };
