// @shopify/polaris/dist/esm/components/CheckableButton/CheckableButton.js
import React$1, { useRef, useContext, useEffect } from 'react';
import { classNames } from '../../utilities/css.js';
import { Checkbox as Checkbox$1 } from '../Checkbox/Checkbox.js';
import { ResourceListContext } from '../../utilities/resource-list/context.js';
import styles from './CheckableButton.scss.js';

function CheckableButton({
  accessibilityLabel,
  label = '',
  onToggleAll,
  selected,
  selectMode,
  plain,
  measuring,
  disabled,
  smallScreen
}) {
  const checkBoxRef = useRef(null);
  const {
    registerCheckableButtons
  } = useContext(ResourceListContext);
  let currentKey = 'bulkLg';

  if (plain) {
    currentKey = 'plain';
  } else if (smallScreen) {
    currentKey = 'bulkSm';
  }

  useEffect(() => {
    if (checkBoxRef.current && registerCheckableButtons) {
      registerCheckableButtons(currentKey, checkBoxRef.current);
    }
  }, [currentKey, registerCheckableButtons]);
  const className = plain ? classNames(styles.CheckableButton, styles['CheckableButton-plain']) : classNames(styles.CheckableButton, selectMode && styles['CheckableButton-selectMode'], selected && styles['CheckableButton-selected'], measuring && styles['CheckableButton-measuring']);
  return /*#__PURE__*/React$1.createElement("div", {
    className: className,
    onClick: onToggleAll
  }, /*#__PURE__*/React$1.createElement("div", {
    className: styles.Checkbox
  }, /*#__PURE__*/React$1.createElement(Checkbox$1, {
    label: accessibilityLabel,
    labelHidden: true,
    checked: selected,
    disabled: disabled,
    onChange: onToggleAll,
    ref: checkBoxRef
  })), /*#__PURE__*/React$1.createElement("span", {
    className: styles.Label
  }, label));
}

export { CheckableButton };
