// @shopify/polaris/dist/esm/components/ActionMenu/components/RollupActions/RollupActions.js
import React$1 from 'react';
import { useI18n } from '../../../../utilities/i18n/hooks.js';
import { HorizontalDotsMinor } from '@shopify/polaris-icons';
import { Popover as Popover$1 } from '../../../Popover/Popover.js';
import { ActionList as ActionList$1 } from '../../../ActionList/ActionList.js';
import { Button as Button$1 } from '../../../Button/Button.js';
import { useToggle as useToggle$1 } from '../../../../utilities/use-toggle.js';
import styles from './RollupActions.scss.js';

function RollupActions({
  items = [],
  sections = []
}) {
  const i18n = useI18n();
  const {
    value: rollupOpen,
    toggle: toggleRollupOpen
  } = useToggle$1(false);

  if (items.length === 0 && sections.length === 0) {
    return null;
  }

  const activatorMarkup = /*#__PURE__*/React$1.createElement("div", {
    className: styles.RollupActivator
  }, /*#__PURE__*/React$1.createElement(Button$1, {
    outline: true,
    icon: HorizontalDotsMinor,
    accessibilityLabel: i18n.translate('Polaris.ActionMenu.RollupActions.rollupButton'),
    onClick: toggleRollupOpen
  }));
  return /*#__PURE__*/React$1.createElement(Popover$1, {
    active: rollupOpen,
    activator: activatorMarkup,
    preferredAlignment: "right",
    onClose: toggleRollupOpen,
    hideOnPrint: true
  }, /*#__PURE__*/React$1.createElement(ActionList$1, {
    items: items,
    sections: sections,
    onActionAnyItem: toggleRollupOpen
  }));
}

export { RollupActions };
