// @shopify/polaris/dist/esm/components/TopBar/components/Search/Search.js
import React$1 from 'react';
import { ThemeProvider as ThemeProvider$1 } from '../../../ThemeProvider/ThemeProvider.js';
import { classNames } from '../../../../utilities/css.js';
import { SearchDismissOverlay as SearchDismissOverlay$1 } from '../SearchDismissOverlay/SearchDismissOverlay.js';
import styles from './Search.scss.js';

function Search({
  visible,
  children,
  onDismiss,
  overlayVisible = false
}) {
  if (children == null) {
    return null;
  }

  const overlayMarkup = visible ? /*#__PURE__*/React$1.createElement(SearchDismissOverlay$1, {
    onDismiss: onDismiss,
    visible: overlayVisible
  }) : null;
  return /*#__PURE__*/React$1.createElement(React$1.Fragment, null, overlayMarkup, /*#__PURE__*/React$1.createElement("div", {
    className: classNames(styles.Search, visible && styles.visible)
  }, /*#__PURE__*/React$1.createElement(ThemeProvider$1, {
    theme: {
      colorScheme: 'dark'
    }
  }, /*#__PURE__*/React$1.createElement("div", {
    className: styles.SearchContent
  }, /*#__PURE__*/React$1.createElement("div", {
    className: styles.Results
  }, children)))));
}

export { Search };
