// @shopify/polaris/dist/esm/components/TopBar/components/Menu/components/Message/Message.js
import React$1 from 'react';
import { Popover as Popover$1 } from '../../../../../Popover/Popover.js';
import { Badge as Badge$1 } from '../../../../../Badge/Badge.js';
import { Button as Button$1 } from '../../../../../Button/Button.js';
import { Stack as Stack$1 } from '../../../../../Stack/Stack.js';
import { Heading as Heading$1 } from '../../../../../Heading/Heading.js';
import { TextContainer as TextContainer$1 } from '../../../../../TextContainer/TextContainer.js';
import { Link as Link$1 } from '../../../../../Link/Link.js';
import styles from './Message.scss.js';

function Message({
  title,
  description,
  action,
  link,
  badge
}) {
  const badgeMarkup = badge && /*#__PURE__*/React$1.createElement(Badge$1, {
    status: badge.status
  }, badge.content);
  const {
    to,
    content: linkContent
  } = link;
  const {
    onClick,
    content: actionContent
  } = action;
  return /*#__PURE__*/React$1.createElement("div", {
    className: styles.Section
  }, /*#__PURE__*/React$1.createElement(Popover$1.Section, null, /*#__PURE__*/React$1.createElement(Stack$1, {
    vertical: true,
    spacing: "tight"
  }, /*#__PURE__*/React$1.createElement(TextContainer$1, null, /*#__PURE__*/React$1.createElement(Heading$1, null, title, badgeMarkup), /*#__PURE__*/React$1.createElement("p", null, description)), /*#__PURE__*/React$1.createElement(Link$1, {
    url: to
  }, linkContent), /*#__PURE__*/React$1.createElement(Button$1, {
    plain: true,
    onClick: onClick
  }, actionContent))));
}

export { Message };
